File: EmojibleIOS/EmojibleIOS/DataSources/UserDataSource.swift
//
//  UserDataSource.swift
//  EmojibleIOS
//
//  Created by Haldun Balim on 2.11.2020.
//

import Foundation
import Firebase

class UserDataSource{
    let database = Firestore.firestore()
    var snapshotListener:ListenerRegistration?
    let notificationCenter = NotificationCenter.default
    
    private init(){
    }
    private static let instance = UserDataSource()
    public static func getInstance() -> UserDataSource{
        return .instance
    }
    
    func getCurrentUserInfo(completion: @escaping (UserModel?) -> Void){
        if let currentUser = AuthenticationManager.getInstance().currentUser{
            let userDocRef = database.collection("Users").document(currentUser.uid)
            userDocRef.getDocument { (document, error) in
                if let document = document, document.exists {
                    guard let dict = document.data() else { return }
                    let userInfo = UserFactory.getInstance().create(dict: dict)
                    completion(userInfo)
                }else{
                    completion(nil)
                }
            }
        }else{
            completion(nil)
        }
    }
    
    func startObservingUserModel(){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        let userDocRef = database.collection("Users").document(currentUser.uid)
        snapshotListener = userDocRef.addSnapshotListener{ [unowned self] documentSnapshot, error in
            guard let document = documentSnapshot else { return }
            guard let dict = document.data() else { return }
            notificationCenter.post(name: .userModelChanged, object: nil, userInfo:["userModel":UserFactory.getInstance().create(dict: dict)])
        }
    }
    
    func stopObservingUserModel(){
        guard let snapshotListener = snapshotListener else { return }
        snapshotListener.remove()        
    }
    
    public func writeUserData(user:UserModel){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        let uid = currentUser.uid
        self.database.collection("Users").document(uid).setData(user.dictionary)
    }
    
    public func editUserData(newData: [String:Any]){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        let uid = currentUser.uid
        self.database.collection("Users").document(uid).updateData(newData)
    }
    
    public func resetUserClassId(classId: String){
        database.collection("Users").whereField("classId", isEqualTo: classId).getDocuments(){ (querySnapshot, err) in
            if let err = err {
                print("Error getting documents: \(err)")
            }
            else{
                for document in querySnapshot!.documents{
                    self.database.collection("Users").document(document.documentID).updateData(["classId" : ""])
                }
            }
        }
    }
}

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/dataSources/UserDataSource.kt
package com.dji.emojibleandroid.dataSources

import android.os.Build
import androidx.annotation.RequiresApi
import com.dji.emojibleandroid.models.UserModel
import com.dji.emojibleandroid.models.modelFactories.UserFactory
import com.dji.emojibleandroid.services.AuthenticationManager
import com.dji.emojibleandroid.services.Changes
import com.dji.emojibleandroid.services.NotificationCenter
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import java.lang.Exception

class UserDataSource private constructor(){
    private val database = Firebase.firestore
    var snapShotListener: ListenerRegistration? = null
    private val notificationCenter = NotificationCenter.instance

    @RequiresApi(Build.VERSION_CODES.O)
    fun getCurrentUserInfo(completion: (UserModel?) -> Unit) {
        val currentUser = AuthenticationManager.instance.currentUser
        currentUser?.let {
            val userDocRef = database.collection("Users").document(it.uid)
            userDocRef.get().addOnCompleteListener { document ->
                if (document.isSuccessful) {
                    if (document.result!!.exists()) {
                        val dict = document.result!!.data
                        val userInfo = UserFactory.instance.create((dict as HashMap<String, Any>))
                        completion(userInfo)
                    }
                    completion(null)
                } else {
                    completion(null)
                }
            }
        }
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun startObservingUserModel() {
        val currentUser = AuthenticationManager.instance.currentUser
        val userDocRef = currentUser?.let { database.collection("Users").document(it.uid) }
        snapShotListener = userDocRef?.addSnapshotListener { documentSnapshot, error ->
            if (documentSnapshot == null) {
                return@addSnapshotListener
            }
            val dict = documentSnapshot.data ?: return@addSnapshotListener

            notificationCenter.post(
                Changes.userModelChagend,
                null,
                "userModel" to UserFactory.instance.create(dict as HashMap<String, Any>)
            )
        }
    }

    fun stopObservingUserModel() {
        snapShotListener?.remove()
    }

    fun writeData(user: UserModel) {
        val currentUser = AuthenticationManager.instance.currentUser
        currentUser?.uid?.let { database.collection("Users").document(it).set(user.dictionary) }
    }

    fun editUserData(newData: HashMap<String, Any?>) {
        val currentUser = AuthenticationManager.instance.currentUser ?: return
        val uid = currentUser.uid
        database.collection("Users").document(uid).update(newData)
    }

    fun resetUserClassId() {
        val currentUser = AuthenticationManager.instance.currentUser ?: return
        val uid = currentUser.uid
        database.collection("Users").document(uid).update(mapOf<String, String>("classId" to ""))
    }

    fun resetUserClassIdForAllUsers(classId: String) {
        database.collection("Users").whereEqualTo("classId", classId).get().addOnCompleteListener {
            if (it.isSuccessful) {
                for (document in it.result?.documents!!) {
                    database.collection("Users").document(document.id).update(mapOf<String, String>("classId" to ""))
                }
            } else if (it.isCanceled) {
                throw Exception("Error getting documents ${it.exception?.message}")
            }
        }
    }

    private object HOLDER {
        val INSTANCE = UserDataSource()
    }

    companion object {
        val instance: UserDataSource by lazy { HOLDER.INSTANCE }
    }
}